tools/classes/PresenceCompiler.ts

```diff
@@ -0,0 +1,296 @@
+import { execSync } from "node:child_process";
+import { existsSync, rmSync, writeFileSync } from "node:fs";
+import { createRequire } from "node:module";
+import { basename, dirname, resolve } from "node:path";
+import { fileURLToPath } from "node:url";
+
+import actions from "@actions/core";
+import chalk from "chalk";
+import { config } from "dotenv";
+import webpack from "webpack";
+
+import { ErrorInfo } from "ts-loader/dist/interfaces";
+import { getFolderLetter } from "../util.js";
+
+const require = createRequire(import.meta.url),
+	rootPath = resolve(fileURLToPath(new URL(".", import.meta.url)), "../..");
+
+if (!process.env.GITHUB_ACTIONS) config({ path: resolve(rootPath, ".env") });
+
+export const tsconfig = JSON.stringify({
+	extends: "../../../tsconfig.json",
+});
+
+export default class PresenceCompiler {
+	cwd: string;
+
+	constructor(
+		public options?: {
+			cwd?: string;
+			webpack?: webpack.Configuration;
+		}
+	) {
+		this.cwd = options?.cwd ?? rootPath;
+	}
+
+	getPresenceFolder(presence: string) {
+		return resolve(this.cwd, "websites", getFolderLetter(presence), presence);
+	}
+
+	async compilePresence(
+		presences: string | string[],
+		options: {
+			output?: string;
+			transpileOnly?: boolean;
+			emit?: boolean;
+		} = {}
+	) {
+		options.emit ??= true;
+		options.transpileOnly ??= false;
+
+		const webpackConfig: webpack.Configuration = {
+			mode: "production",
+			devtool: "inline-source-map",
+			resolve: {
+				extensions: [".ts"],
+			},
+
+			output: options.emit
+				? {
+						iife: false,
+						path: options.output,
+						filename: "[name].js",
+				  }
+				: undefined,
+			module: {
+				rules: [
+					{
+						test: /\.ts$/,
+						loader: "ts-loader",
+						exclude: /node_modules/,
+						options: {
+							transpileOnly: options.transpileOnly,
+							errorFormatter: (error: ErrorInfo) => {
+								actions.error(chalk.redBright(error.content), {
```

Comment:
I believe error would give it a red box anyway
```suggestion
								actions.error(error.content, {
```